operator/src/main/resources/application.properties

```diff
@@ -16,6 +16,12 @@ kc.operator.keycloak.resources.limits.memory=2Gi
 # https://quarkus.io/guides/deploying-to-kubernetes#environment-variables-from-keyvalue-pairs
 quarkus.kubernetes.env.vars.related-image-keycloak=${kc.operator.keycloak.image}
 
+#https://quarkus.io/guides/deploying-to-kubernetes#container-resources-management
+quarkus.kubernetes.resources.requests.memory=450Mi
+quarkus.kubernetes.resources.requests.cpu=300m
+quarkus.kubernetes.resources.limits.memory=450Mi
+quarkus.kubernetes.resources.limits.cpu=300m
```

Comment:
I'm kinda skeptical about the CPU limits as they provide more cons than pros. As the CPU is a compressible resource in k8s, the pod that reached the CPU limit is not evicted from a node, but rather a CPU is throttled and performance significantly degraded - I'd like to avoid it. That means even when there are available CPU resources on the node, the pod is restricted.

When we do **NOT** set the CPU limit, the pod can use more resources when needed, and it is not restricted at all. As the Completely Fair Scheduler (CFS) is used in k8s, the CPU time is fairly distributed among processes/pods. It prevents the pod, without the CPU limit, from grabbing all the available CPU time and leaving all other pods on the node starved. 

Thus, I'd suggest not setting the CPU limit.

@Pepo48 @vmuzikar @shawkins @ahus1 WDYT?

Reply:
Based on the [best practices for the Operators](https://master.sdk.operatorframework.io/docs/best-practices/managing-resources/#general-guidelines), CPU limits are optional but still recommended:
> What happens when the resource limits are not set?
> ...
> when there is CPU contention on the node, neighbour pods may get slowed down to the CPU cores they requested

Reply:
I would also vote for having a limit. If 300m is giving us a reasonable startup time and has padding beyond what is nominally needed to manage 10 keycloaks, then that seems reasonable. Having the limit from a resource planning perspective is always preferred.

Reply:
@mabartos thanks for sharing this. I wondered actually, why we don't set the cpu limits for Keycloak containers, while we have memory limits already implemented. That's partially why I got confused and [implemented](https://github.com/keycloak/keycloak/compare/main...Pepo48:keycloak:kc-cpu-limits) it at first (the numbers are made-up placeholders). You kinda explained why. I already created an issue for it [here](https://github.com/keycloak/keycloak/issues/30860).

A little observation - with a larger number of Keycloaks provisioned at once you're more likely to hit the memory [request](https://github.com/keycloak/keycloak/blob/ee58ce6f51b25003afb5d59625ef76f9a4f93c5e/operator/src/main/resources/application.properties#L13) imposed to the Keycloak containers themselves. During the testing I often ended up in the state where scheduler could not continue with the provisioning due to not meeting these requirements. I know it depends on the amount of resources, but I observed the same with multiple environments (remotely and locally). The operator was safe - just small increments small depending on the number of managed operands.

Reply:
Based on an offline discussion, we decided to keep a CPU limit on the Operator Pod. Question is to what value we should set it to. @mabartos suggested increasing it above 300m.

Reply:
@Pepo48 @vmuzikar @shawkins 

After the private discussions, I agree to have the CPU limit for the Operator as the resource utilization is not so dynamic and kinda stable over time. As I've mentioned in these discussions, I'd like to see a probably higher CPU limit than the suggested one (300m). At first sight, it seems to me really low to limit the Operator on such values as the startup for the Operator might be more resource-consuming. 

As stated above, the CPU is compressible and does not require to have such a small difference between requests/limits values. What about setting the CPU limit for the Operator to, at least, ~**700-800m**? It is not much either, but it can address possible peaks for the Operator. 

WDYT?

Reply:
As I have been asked above: 

I agree it is a best practice to limit the CPU usage for apps, to a lesser degree also for Operators as they are usually idle. So I would support both approaches. I see you agreed to use a CPU limit, which is IMHO fine. 

I had a look at our nightly deployment, and an operator that is idle for a single Keycloak instance, I see 5-15m CPU once it is running for about 10 minutes, and up to 20m during the first 10 min. So 300m CPU as a request should work well, even when managing several Keycloaks - maybe lowering it to 100m would be possible as well, if limit is still at 300 or even a larger value as @mabartos suggested.

Reply:
That matches my observations. To be clear, I determined 300m as the upper limit (with 30% margin already included). I deployed 10 Keycloaks + 10 realm imports, yet simple configurations. The highest consumption CPU observed was around 220m.

The 300m request just matches the 300m limit, simply because I thought the same or close values are preferable like with the memory (otherwise I would ended up with something like 120m as a request). But I think it was Martin, who pointed out earlier today that it's fine when values differ for the CPU.

Anyway, let's be rather safe than sorry - 300/700 it is. 😊